Add tsh proxy types aws,db,ssh to CLI ref (#17716)
gravitational/teleport: docs/pages/reference/cli.mdx

$ tsh play --format=json ~/play/0c0b81ed-91a9-4a2a-8d7c-7495891a6ca0.tar | jq '.event
```

### tsh proxy db

> Start local TLS proxy for database connections when using Teleport in single-port mode.

Reviewer: ```
$ grep -Rn "single-port mode" docs/pages
docs/pages/reference/cli.mdx:332:Start local TLS proxy for database connections when using Teleport in single-port mode.
docs/pages/reference/cli.mdx:379:Start local TLS proxy for `ssh` connections when using Teleport in single-port mode.
```
Since "single-port mode" is only used on this page, is there another term we can use to be more consistent with other parts of the docs?

Reviewer 2: _TLS Routing_ is the right term to use here IMO

> Database connections can be made with certificates through database clients or tunnel mode

Reviewer: I think we should use the active voice here—instead of saying "can be made", we should clarify what is making the connections.

> that does not require authentication.

Reviewer: ```suggestion
, which does not require authentication.
```

```code
$ tsh proxy db [<flags>] <db>
```

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |

> | `--cert-file` | none | Certificate file| Certificate file for proxy client TLS configuration |

Reviewer: Is this the content of the certificate file or the path?

> | `--key-file` | none | Key file| Key file for proxy client TLS configuration |

Reviewer: Same comment as above re: "Key file"

> | `tunnel` | none | Enable tunnel mode | Open authenticated tunnel using database's client certificate so clients don't need to authenticate |

Reviewer: ```suggestion
| `tunnel` | none | Enable tunnel mode | Open authenticated tunnel using a database's client certificate so clients don't need to authenticate |
```

> | `--db-user` | none | Database user| Optional database user to log in as. |

Reviewer: ```suggestion
| `--db-user` | none | Database user| Optional database user to log in as |
```
For consistent punctuation in this table column

Use the following command to connect to the database:
  $ mysql --port 49415 --host localhost --protocol TCP
```

### tsh proxy ssh

Start local TLS proxy for `ssh` connections when using Teleport in single-port mode.
This is typically used as part of the SSH client configuration to use `ssh` as a client 
through Teleport.  See [OpenSSH Guide](../server-access/guides/openssh.mdx)
on configuring OpenSSH servers and clients.  The `tsh config` output will include `tsh proxy ssh`
within the `ProxyCommand`.

```code
$ tsh proxy ssh [<flags>] <[user@]host>
```

#### Arguments
- `<[user@]host>` Remote hostname and the login to use

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |
| `--cluster` | none | Teleport Cluster | Specify the Teleport cluster to connect |

#### [Global Flags](#tsh-global-flags)

These flags are available for all commands `--login, --proxy, --user, --ttl, --identity, --cert-format, --insecure, --auth, --skip-version-check, --debug, --jumphost, --format`.
Run `tsh help <subcommand>` or see the [Global Flags Section](#tsh-global-flags)

#### Examples

> ```code

Reviewer: Is there ever a reason to use this command outside of the `tsh config` output? That might be good to include here.

$ tsh play --format=json ~/play/0c0b81ed-91a9-4a2a-8d7c-7495891a6ca0.tar | jq '.event
```

### tsh proxy db

Start a local TLS proxy for database connections when using Teleport with TLS
Routing enabled. Clients can connect to a Teleport-registered database through
the local proxy.

```code
$ tsh proxy db [<flags>] <db>
```

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |
| `--cert-file` | none | string | Path to the certificate file for configuring TLS on the local proxy.|
| `--cluster` | none | string | The name of the Teleport cluster to connect to. |
| `--key-file` | none | string | Path to the private key file for configuring TLS on the local proxy.|
| `--db-name` | none | string | Optional database name to log in to. |
| `--db-user` | none | string | Optional database user to log in as. |

> | `--port` | string | Source port used by the local proxy.|

Reviewer 3: This line is missing the default value column. I assume is should also be `none`?

| `--tunnel` | none | Boolean | Open an authenticated tunnel using a database's client certificate so clients don't need to authenticate. |

#### [Global Flags](#tsh-global-flags)

These flags are available for all commands `--login, --proxy, --user, --ttl, --identity, --cert-format, --insecure, --auth, --skip-version-check, --debug, --jumphost, --format`.
Run `tsh help <subcommand>` or see the [Global Flags Section](#tsh-global-flags)

#### Examples
```code
$ tsh proxy db <db>

# Proxy a connection to mysql db on local port 10700
$ tsh proxy db --port 10700  mysql-db
# Started DB proxy on 127.0.0.1:10700

# Use following credentials to connect to the mysql-db proxy:
#  ca_file=/Users/jeff/.tsh/keys/teleport.example.com/cas/teleport.example.com.pem
#  cert_file=/Users/jeff/.tsh/keys/teleport.example.com/jeff-db/tele1c/mysql-db-x509.pem
#  key_file=/Users/jeff/.tsh/keys/teleport.example.com/jeff

# Proxy a connection to mysql db with no credentials required
$ tsh proxy db --tunnel mysql-db
Started authenticated tunnel for the MySQL database "mysql-db" in cluster "teleport.example.com" on 127.0.0.1:49415.

Use the following command to connect to the database:

>   $ mysql --port 49415 --host localhost --protocol TCP

Reviewer 3: Looks like there's a mix of docs-as-comments and regular un-commented copy(L434) in this code block. Would it make sense to elevate all copy and have separate code blocks?

Author: > 

> Changing my review status back. I incorrectly assumed you addressed all comments in my previous review with your last set of changes.

@alexfornuto should be all addressed now.  I turned off auto-merge

```

### tsh proxy ssh

Start a local TLS proxy for `ssh` connections when using Teleport in TLS Routing mode.
This is typically used as part of the SSH client configuration to use `ssh` as a client 

> through Teleport.  See the [OpenSSH Guide](../server-access/guides/openssh.mdx)

Reviewer 3: ```suggestion
through Teleport.  See the [OpenSSH Guide](../server-access/guides/openssh.mdx) guide
```

or doc, page, what have you.

| - | - | - | - |
| `--cluster` | none | Teleport Cluster | The name of the Teleport cluster to connect to.|

#### [Global Flags](#tsh-global-flags)

These flags are available for all commands `--login, --proxy, --user, --ttl, --identity, --cert-format, --insecure, --auth, --skip-version-check, --debug, --jumphost, --format`.
Run `tsh help <subcommand>` or see the [Global Flags Section](#tsh-global-flags)

#### Examples

Run the following command to generate an OpenSSH client configuration:

```code
$ tsh config
```

This command produces the following configuration: 

```
# Common flags for all example.com hosts
Host *.example.com example.com
    UserKnownHostsFile "/Users/jeff/.tsh/known_hosts"
    IdentityFile "/Users/jeff/.tsh/keys/enterprise.teleportdemo.com/jeff"
    CertificateFile "/Users/jeff/.tsh/keys/example.com/jeff-ssh/example.com-cert.pub"
    PubkeyAcceptedKeyTypes +ssh-rsa-cert-v01@openssh.com
    HostKeyAlgorithms ssh-rsa-cert-v01@openssh.com

# Flags for all example.com hosts except the proxy
Host *.example.com !example.com
    Port 3022
    ProxyCommand "/usr/local/bin/tsh" proxy ssh --cluster=example.com --proxy=example.com %r@%h:%p
```

> 

Reviewer 3: Should this area include a paragraph about where this config goes? "Typically `~/.ssh/config` for macOS and Linux. XYZ for PuTTY...", something to that effect?

Reviewer 3: > > 
> 
> > Changing my review status back. I incorrectly assumed you addressed all comments in my previous review with your last set of changes.
> 
> @alexfornuto should be all addressed now. I turned off auto-merge

I'm not saying this is a required change or anything like that, just wanna make sure it wasn't missed.

Author: Thanks, yes I did put something in

the local proxy.

```code
$ tsh proxy db [<flags>] <db>
```

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |
| `--cert-file` | none | string | Path to the certificate file for configuring TLS on the local proxy.|
| `--cluster` | none | string | The name of the Teleport cluster to connect to. |
| `--key-file` | none | string | Path to the private key file for configuring TLS on the local proxy.|
| `--db-name` | none | string | Optional database name to log in to. |
| `--db-user` | none | string | Optional database user to log in as. |
| `--port` | none | string | Source port used by the local proxy.|
| `--tunnel` | none | Boolean | Open an authenticated tunnel using a database's client certificate so clients don't need to authenticate. |

#### [Global Flags](#tsh-global-flags)

These flags are available for all commands `--login, --proxy, --user, --ttl, --identity, --cert-format, --insecure, --auth, --skip-version-check, --debug, --jumphost, --format`.
Run `tsh help <subcommand>` or see the [Global Flags Section](#tsh-global-flags)

#### Examples

Proxy a DB connection to a named database
```code
$ tsh proxy db <db>
```

> Proxy a connection to mysql db on local port 10700

Reviewer 3: ```suggestion
Proxy a connection to mysql db on local port 10700:
```

```code
$ tsh proxy db --port 10700  mysql-db
# Started DB proxy on 127.0.0.1:10700

# Use following credentials to connect to the mysql-db proxy:
#  ca_file=/Users/jeff/.tsh/keys/teleport.example.com/cas/teleport.example.com.pem
#  cert_file=/Users/jeff/.tsh/keys/teleport.example.com/jeff-db/tele1c/mysql-db-x509.pem
#  key_file=/Users/jeff/.tsh/keys/teleport.example.com/jeff
```

> Proxy a connection to mysql db with no credentials required

Reviewer 3: ```suggestion
Proxy a connection to mysql db with no credentials required:
```

$ curl http://127.0.0.1:10700/api/users
```

### tsh proxy aws

> Start local proxy for AWS access.  The user should already be logged into

Reviewer: ```suggestion
Start a local proxy for AWS access.  The user should already be logged into
```

at least one AWS application.

```code
$ tsh proxy aws [<flags>] 
```

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |

> | `--app` | Currently logged in AWS app | AWS app| Optional Name of the AWS application to use if logged into multiple |

Reviewer: ```suggestion
| `--app` | Currently logged in AWS app | AWS app| Optional name of the AWS application to use if logged into multiple |
```

> | `--app` | Currently logged in AWS app | AWS app| Optional Name of the AWS application to use if logged into multiple |

Reviewer: I'm not sure what an example of an allowed value would be here. Are there any restrictions on formatting of the application name? What would be the source of truth for the application name?

Reviewer 3: @ptgott can we mark this as resolved now?

| `-p, --port` | none | port number | Specify the source port for the local proxy |

> | ` -e, --endpoint-url` | HTTP Proxy | Endpoint URL | Run local proxy to serve as an AWS endpoint URL. If not specified, local proxy serves as an HTTPS proxy. |

Reviewer: ```suggestion
| ` -e, --endpoint-url` | HTTP Proxy | Endpoint URL | Run the local proxy to serve as an AWS endpoint URL. If not specified, local proxy serves as an HTTPS proxy. |
```

> | `-f, --format` | unix | AWS app| Optional Name of the AWS application to use if logged into multiple |

Reviewer: Double-checking whether the final two columns of this row should be the same as those in the first row

These flags are available for all commands `--login, --proxy, --user, --ttl, --identity, --cert-format, --insecure, --auth, --skip-version-check, --debug, --jumphost, --format`.
Run `tsh help <subcommand>` or see the [Global Flags Section](#tsh-global-flags)

### tsh proxy aws

> Start a local proxy for AWS access. The user must already be logged into at

Reviewer 3: ```suggestion
Start a local proxy for AWS access. The user must already be logged in to at
```

[reference](https://www.grammarly.com/blog/into-vs-in-to/#:~:text=I%20dropped%20into,to%20say%20hello.)

least one AWS application via Teleport before the proxy can start.

```code
$ tsh proxy aws [<flags>] 
```

#### Flags

| Name | Default Value(s) | Allowed Value(s) | Description |
| - | - | - | - |

> | `--app` | Currently logged in AWS app | string | Optional name of the AWS application (as shown in `tsh apps ls`) to use if logged into multiple |

Reviewer 3: ```suggestion
| `--app` | Currently logged in AWS app | string | Optional name of the AWS application (as shown in `tsh apps ls`) to use if logged in to multiple |
```

| `-p, --port` | none | port number | Specify the source port for the local proxy |
| ` -e, --endpoint-url` | HTTP Proxy | Endpoint URL | Run the local proxy to serve as an AWS endpoint URL. If not specified, the local proxy serves as an HTTPS proxy. |

> | `-f, --format` | unix | `text`, `unix`, `command-prompt`, or `powershell` | Optional format for printing environmen  variables for the AWS proxy |

Reviewer 3: ```suggestion
| `-f, --format` | unix | `text`, `unix`, `command-prompt`, or `powershell` | Optional format for printing environment variables for the AWS proxy |
```
